core/sis-feature/src/main/java/org/apache/sis/filter/Optimization.java

```diff
@@ -0,0 +1,297 @@
+    /**
+     * An arbitrary object meaning that a filter or expression optimization in under progress.
+     */
+    private static final Object COMPUTING = Void.TYPE;
+
+    /**
+     * Filters and expressions already optimized. Also used for avoiding never-ending loops.
+     * The map is created when first needed.
+     *
+     * <div class="note"><b>Note:</b> the same map is used for filters and expressions.
+     * It is not a problem if keys do not implement the two interfaces in same time.
+     * If it happens anyway, it should still be okay because the method signatures are
+     * the same in both interfaces (only the return type changes), so the same methods
+     * would be invoked no matter if we consider the keys as a filter or an expression.</div>
+     */
+    private Map<Object,Object> done;
+
+    /**
+     * Creates a new instance.
+     */
+    public Optimization() {
+    }
+
+    /**
+     * Optimizes or simplifies the given filter. If the given instance implements the {@link OnFilter} interface,
+     * then its {@code optimize(this)} method is invoked. Otherwise this method returns the given filter as-is.
+     *
+     * @param  <R>     the type of resources (e.g. {@code Feature}) used as inputs.
+     * @param  filter  the filter to optimize, or {@code null}.
+     * @return the optimized filter, or {@code null} if the given filter was null.
+     *         May be {@code filter} if no optimization or simplification has been applied.
+     * @throws IllegalArgumentException if the given filter is already in process of being optimized
+     *         (i.e. there is a recursive call to {@code apply(…)} for the same filter).
+     */
+    public <R> Filter<? super R> apply(final Filter<R> filter) {
+        if (!(filter instanceof OnFilter<?>)) {
+            return filter;
+        }
+        final boolean isFirstCall = (done == null);
+        if (isFirstCall) {
+            done = new IdentityHashMap<>();
+        }
+        try {
+            final Object previous = done.putIfAbsent(filter, COMPUTING);
+            if (previous == COMPUTING) {
+                throw new IllegalArgumentException(Errors.format(Errors.Keys.RecursiveCreateCallForKey_1, filter.getOperatorType()));
+            }
+            @SuppressWarnings("unchecked")
+            Filter<? super R> result = (Filter<? super R>) previous;
+            if (result == null) {
+                result = ((OnFilter<R>) filter).optimize(this);
+                if (done.put(filter, result) != COMPUTING) {
+                    // Should not happen unless this `Optimization` is used concurrently in many threads.
+                    throw new ConcurrentModificationException();
+                }
+            }
+            return result;
+        } finally {
+            if (isFirstCall) {
+                done = null;
+            }
+        }
+    }
+
+    /**
+     * Filter than can be optimized. Each filter implementation knows which rules can be applied.
```

Comment:
s/than can be/that can be

Reply:
Fixed 10 occurrences of this error in the code base.